docs/rfcs/749-unify-path-behavior.md

```diff
@@ -0,0 +1,88 @@
+
+Previous Discussion:
+
+- [Specify the behavior of Path](https://forum.beyondstorage.io/t/topic/195)
+- [OS-specific path separator problem](https://github.com/beyondstorage/go-service-ftp/issues/22)
+
+## Background
+
+Unix and Windows all use a different syntax for filesystem paths. Normally, when specifying file paths on Windows, we would use backslashes(`\`). However, in Java, and many other places outside the Windows world, backslashes are the escape character, so we have to double them up, like `C:\\trash\\blah\\blah`. Forward slashes(`/`), on the other hand, do not need to be doubled up and work on both Windows and Unix. There is no harm in having double forward slashes. They do nothing to the path and just take up space (`//` is equivalent to `/./`).
+
+In our services, we have to handle paths like path splicing and path conversion. And here comes the problems:
+
+- Different path separators may exist in a path at the same time, is it possible or how to use cross-platform paths for users?
+- For fs storage service on Windows platform, the behavior of `work_dir` is undefined. Drive letters in connection string cannot be handled for Windows at this stage.
+
+## Proposal
+
+### Absolute Path and Relative Path
+
+All our services should support two kinds of path:
+
+- Absolute Path:
+  - For Unix, the absolute path starts with `/`.
+  - For Windows, the absolute path starts with a drive letter, `/` is the current drive.
+- Relative Path: 
+  - The relative path is for the working directory `WorkDir`.
+  - Path with prefix `./` or `../` is allowed.
+
+### WorkDir and path
+
+`WorkDir` specifies the working directory of the process.
+
+- For file system, `WorkDir` SHOULD be a passed in absolute path or relative to the current directory of the process.
+  - The default value is `/`, that means the working directory is the root path for Unix, and the current drive of the process for Windows.
+  - Services should set `WorkDir` to the path name after the evaluation of any symbolic links internal.
+- For object storage, `WorkDir` is the simulated directory or prefix of the object key. 
+  - The default value is `""`.
+  - `WorkDir` SHOULD be the unix style and SHOULD NOT with prefix `/` when the value is not empty.
+  
+`path` is the file path for file system, or an object key for object storage. Also, it could be a prefix filter for `List` operation.
+
+- `path` could be an absolute path or a relative path.
+- Services SHOULD convert it to the absolute path at the beginning of the operation.
+- For the unique key `Object.ID` in storage, it should be an absolute path, unless there's a returned unique identifier like in dropbox.
+- Users SHOULD follow the file and object naming of different services.
+
+### Path Separator
+
+All the passed in path SHOULD be unix style, no matter Linux platform or Windows platform, no matter object storage service or file system.
+When the drive letter is included for Windows, it should be something like `c:/a/b`.

```

Comment:
Maybe we only need to allow this case inside `go-service-fs`?

Reply:
I found it hard to handle, maybe we can limit `go-service-fs` only support current volume? So that we don't need to support `read("c:\\x\\y\\z")`?

Reply:
> I found it hard to handle, maybe we can limit `go-service-fs` only support current volume? So that we don't need to support read("c:\\x\\y\\z")?

Does it mean we don't need to support drive letter for all the paths including `WorkDir`? Or only for `WorkDir`(to point the working directory as the current volume like `c:\`, but we can write to `d:\test`)?